src/selection/highlight/visualSelection.js

```diff
@@ -103,17 +148,70 @@ class VisualSelection extends Selection {
       return this;
     }
 
+    const {
+      from: visualFromCoords,
+      to: visualToCoords,
+    } = this.visualCellRange;
+
     // We may move in two different directions while searching for visible rows and visible columns.
     const incrementByRow = this.getRowSearchDirection(this.visualCellRange);
     const incrementByColumn = this.getColumnSearchDirection(this.visualCellRange);
     const fromRangeVisual = this.findVisibleCoordsInRange(
-      this.visualCellRange.from, this.visualCellRange.to, incrementByRow, incrementByColumn);
+      visualFromCoords,
+      visualToCoords,
+      incrementByRow,
+      incrementByColumn
+    );
     const toRangeVisual = this.findVisibleCoordsInRange(
-      this.visualCellRange.to, this.visualCellRange.from, -incrementByRow, -incrementByColumn);
-
-    // There is no visual start point (and also visual end point) in the range. We are looking for the first visible cell in a broader range.
+      visualToCoords,
+      visualFromCoords,
+      -incrementByRow,
+      -incrementByColumn
+    );
+
+    // There is no visual start point (and also visual end point) in the range.
+    // We are looking for the first visible cell in a broader range.
     if (fromRangeVisual === null) {
-      this.cellRange = null;
+      // For the "header" selection type, find rows and column indexes, which should be
+      // highlighted, although one of the axes is completely hidden.
+      if (this.settings.type === 'header') {
+        const fromRangeVisualRow = this.findVisibleCoordsInRowsRange(
+          visualFromCoords.row,
+          visualToCoords.row,
+          incrementByRow
+        );
+        const toRangeVisualRow = this.findVisibleCoordsInRowsRange(
+          visualToCoords.row,
+          visualFromCoords.row,
+          -incrementByRow
+        );
+        const fromRangeVisualColumn = this.findVisibleCoordsInColumnsRange(
+          visualFromCoords.col,
+          visualToCoords.col,
+          incrementByRow
+        );
+        const toRangeVisualColumn = this.findVisibleCoordsInColumnsRange(
+          visualToCoords.col,
+          visualFromCoords.col,
+          -incrementByRow
+        );
+
+        // If all rows and columns ranges are hidden then remove the cell range.
+        if (fromRangeVisualRow === null && toRangeVisualRow === null &&
+            fromRangeVisualColumn === null && toRangeVisualColumn === null) {
+          this.cellRange = null;
+        } else {
+
+          this.cellRange = new CellRange(
+            new CellCoords(0, 0), // The highlight coordinates are not taken into consideration while highlighting headers.
+            new CellCoords(fromRangeVisualRow, fromRangeVisualColumn),
+            new CellCoords(toRangeVisualRow, toRangeVisualColumn),
+          );
+        }
+
```

Comment:
I have not idea how to separate it. Probably creating some class for the visual header selection isn't the best idea. Maybe just move this part of code to separate method within the class? I think that it will be more clear. WDYT?

Reply:
I've prepared such an improvement within https://github.com/handsontable/handsontable/pull/7053/commits/5d79673268bd05b88f6fff70383d4de16c7b3ea0. I'm not sure about the usage of the `Separately` word but generally it looks much better.